### DISCUSSION TITLE
Image block to have `loading=(eager|lazy)` option

### DISCUSSION THREAD

**Discussion by crs1138** [WordPress/gutenberg#41475] [Ideas]
Would it be possible to implement a new option for any block containing images (or at least for the image block itself) to control the lazy loading? I tried to do that using block's class and filter `wp_img_tag_add_loading_attr` but the class is attached to the `<figure>` wrapper and the filter works with the `<img>` element.

I'm running out of ideas apart implementing some sort of filter and adding an attribute to the core block. Is there another way? Why not to include this as a part of the block with the default value being `lazy`?

**Comment by t-hamano** [comment]
Hi @crs1138,

Sorry for the late reply.

If you want to change the image `loading` attribute in a particular block, how about using the [render_block_{$blockname}](https://developer.wordpress.org/reference/hooks/render_block_this-name/) filter hook? Since WordPress 6.2, you will be able to control the attribute more simply by using the [WP_HTML_Tag_Processor class](https://make.wordpress.org/core/2022/08/19/a-new-system-for-simply-and-reliably-updating-html-attributes/).

Would this approach help?

**Comment by jerclarke** [comment]
Have to say, this is a strange omission. 

I have the following in the `footer.html` of my child theme: 

```html
<!-- wp:image {"linkDestination":"custom"} -->
			<figure class="wp-block-image"><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="<?php echo esc_url( get_stylesheet_directory_uri() ); ?>/assets/images/by-nc-sa-ccLicense.png" alt="CC-BY-NC-SA" loading='lazy'/></a></figure>
			<!-- /wp:image -->
```

Which generates this error in the **Site Editor** when I reset the theme:

<img width="476" height="245" alt="Screenshot 2025-11-29 at 10 10 13" src="https://github.com/user-attachments/assets/d7f0c49c-3cf6-4c03-a03c-86aacb00c3b4" />

Console: 
<img width="1004" height="156" alt="Screenshot 2025-11-29 at 10 08 17" src="https://github.com/user-attachments/assets/1606dad9-12d5-4ca0-aaa3-b7fccc5de51a" />

Am I really not able to add `loading=lazy` to any images built into a FSE theme by default? Seems like a really bad scenario. 

Overall WP should just add `loading=lazy` by default to images like this, or incorporate it into the logic that inserts or omits it depending on "which image it is in the page". At the bare minimum, there should be a flag like `<!-- wp:image {"loading":"lazy"} -->` (as well as the obvious companion `<!-- wp:image {"fetchPriority":"high"} -->`). 

As it stands the only way to eliminate this error without elaborate hook juggling is to have an image in my footer without `loading=lazy`, an optimization crime that will send me straight to jail.

**Reply by admcfajn** [reply]
This might help: https://make.wordpress.org/core/2023/07/13/image-performance-enhancements-in-wordpress-6-3/

**Reply by jerclarke** [reply]
So you'd suggest I add a filter to control this? I've done those filters and they're a pain that normal users will never figure out. Thank you for the pointer though, you're right that it's a workaround solution for the time being. 

Still, I think it should be controllable via. the templates and interface, it's a natural property of an image in 2025, just like the `src` and `alt`.

**Reply by jerclarke** [reply]
Ah ok, it turns out the image block **tries** add the loading=lazy, with or without a filter (I created a filter before realizing it wasn't needed). The problem is my image has no width/height, and thus loading=lazy isn't being applied to it, because the internal addition of loading=lazy is subverted when an image has no width/height....

This code in my template part triggers loading=lazy on the frontend: 

```html
<!-- wp:image {"linkDestination":"custom"} -->
	<figure class="wp-block-image"><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="<?php echo esc_url( get_stylesheet_directory_uri() ); ?>/assets/images/by-nc-sa-ccLicense.png" alt="CC-BY-NC-SA" width="525" height="124" /></a></figure>
<!-- /wp:image -->
```

But now I can't find a way to add a width and height to an image that the Site Editor doesn't choke on. The code above gives the same type of error in the editor: 
<img width="681" height="176" alt="Screenshot 2025-12-09 at 13 59 00" src="https://github.com/user-attachments/assets/2cead559-2574-413b-adff-a071fd485781" />

If I add a width and height in the editor, then copy the template code, I get this:

```html
<!-- wp:image {"width":"525px","height":"124px","scale":"contain","linkDestination":"custom"} -->
<figure class="wp-block-image is-resized"><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="https://jc.test/wp-content/themes/jer-twentytwentyfive-child-theme/assets/images/by-nc-sa-ccLicense.png" alt="CC-BY-NC-SA" style="object-fit:contain;width:525px;height:124px"/></a></figure>
<!-- /wp:image -->
```

The code above has no errors in the editor, but once again I have no loading=lazy on it, because WP refuses to add loading=lazy to an image without `width=` on the tag, the `style=` doesn't affect that logic. The filter doesn't help, because WP already **wants** to add the lazy on there, it just can't because `width=`

So I can't help coming back to the same conclusion: There's no way to insert a specific image in a block theme and give it loading=lazy, because there's no way to insert an image with `width=` without triggering an error in the Site Editor.

**Reply by admcfajn** [reply]
It's a work in progress :) 

To add a specific toggle in the Image block sidebar for `fetchpriority`, we can use the `blocks.registerBlockType` filter to add a custom attribute and then filter the block's `Edit` and `Save` components.

### Step 1: Register the Attribute

```javascript
import { addFilter } from '@wordpress/hooks';

addFilter(
    'blocks.registerBlockType',
    'my-plugin/add-performance-attr',
    ( settings, name ) => {
        if ( name === 'core/image' ) {
            settings.attributes = {
                ...settings.attributes,
                fetchPriority: {
                    type: 'string',
                    default: 'auto',
                },
                loading: {
                    type: 'string',
                    default: 'lazy',
                },
            };
        }
        return settings;
    }
);
```

### Step 2: Add the Sidebar Control

Use `InspectorControls` to add a dropdown in the "Settings" tab of the Image block.

```javascript
import { InspectorControls } from '@wordpress/block-editor';
import { SelectControl, PanelBody } from '@wordpress/components';
import { createHigherOrderComponent } from '@wordpress/compose';

const withFetchPriorityControl = createHigherOrderComponent( ( BlockEdit ) => {
    return ( props ) => {
        if ( props.name !== 'core/image' ) return <BlockEdit { ...props } />;

        const { attributes, setAttributes } = props;

        return (
            <>
                <BlockEdit { ...props } />
                <InspectorControls>
                    <PanelBody title="Performance Settings">
                        <SelectControl
                            label="Fetch Priority"
                            value={ attributes.fetchPriority }
                            options={[
                                { label: 'Auto', value: 'auto' },
                                { label: 'High (LCP Hero)', value: 'high' },
                                { label: 'Low', value: 'low' },
                            ]}
                            onChange={ ( val ) => setAttributes( { fetchPriority: val } ) }
                        />
                        <SelectControl
                          label="Loading"
                          value={attributes.loading}
                          options={[
                            { label: 'Lazy', value: 'lazy' },
                            { label: 'Eager', value: 'eager' },
                          ]}
                          onChange={(val) => setAttributes({ loading: val })}
                        />
                    </PanelBody>
                </InspectorControls>
            </>
        );
    };
}, 'withFetchPriorityControl' );

addFilter( 'editor.BlockEdit', 'core/image', withFetchPriorityControl );
```

### Step 3: Apply to Frontend Render

Instead of filtering the `save` function (which makes blocks "brittle"), use the PHP filter `render_block` to inject the attribute based on the value saved in Gutenberg.

```php
add_filter( 'render_block', function( $block_content, $block ) {
  if ( 'core/image' === $block['blockName'] && ! empty( $block['attrs']['fetchPriority'] ) ) {
    $priority = $block['attrs']['fetchPriority'];
    $loading = $block['attrs']['loading'];
    if ( $priority !== 'auto' ) {
      $block_content = str_replace( '<img ', '<img fetchpriority="' . esc_attr( $priority ) . '" ', $block_content );
    }
    if ( $loading === 'lazy' || $loading === 'eager' ) {
      $block_content = str_replace( '<img ', '<img loading="' . esc_attr( $loading ) . '" ', $block_content );
    }
  }
  return $block_content;
}, 10, 2 );
```

**Reply by jerclarke** [reply]
Wow, that's very impressive, and thank you for working all that code up to demonstrate a solution. 

But am I wrong to think that the more urgent question relates to width/height? I've heard that it's bad for "CLS" to have a lazy loaded image that doesn't have the width and height specified. It seems to me that it's important to have a way to insert images with a width and height, then let WP filter in the fetchpriority/loading attributes automatically (filtering them if necessary). 

I suppose I could use a wave of code like you shared above to add explicit width/height options in the editor to get this effect, but it does strike me as something that ideally would be addressed in core. 

Might it make sense to create a feature request for width/height attributes on the image block?

**Reply by admcfajn** [reply]
> Might it make sense to create a feature request for width/height attributes on the image block?

I don't think so, as those already exist

**Reply by jerclarke** [reply]
So why don't they trigger loading=lazy?

Is the bug then in the automatic loading=lazy code? Should it also be checking for the `style="object-fit:contain;width:525px;height:124px"` code rather than only `width=''`?

If the inline CSS is "good enough" to not worry about CLS, and thus the explicit `width=''` attributes aren't necessary, presumably they're also good enough for loading=lazy.

**Reply by jerclarke** [reply]
Seems like the culprit is this code from `wp_img_tag_add_loading_optimization_attrs()` ([media.php:2192](https://github.com/WordPress/WordPress/blob/fd69b7a993f61f1bb3c3f5ce887ebf97b2fad630/wp-includes/media.php#L2192)), which determines whether an image can qualify for `loading="lazy"` or not: 

```php
	// Images should have dimension attributes for the 'loading' and 'fetchpriority' attributes to be added.
	if ( ! str_contains( $image, ' width="' ) || ! str_contains( $image, ' height="' ) ) {
		return $image;
	}
```

If the current width/height arguments for `wp:image` are sufficient, then that code should accept the resulting inline CSS. 

e.g. If that check used a function like this:
```php
/**
 * Determine if an image string contains width and height properties
 *
 * @param string $image
 * @return bool
 */
function wp_image_has_width_and_height( string $image ): bool {

	if ( str_contains( $image, ' width="' ) && str_contains( $image, ' height="' ) ) {
		return true;
	}

	if ( str_contains( $image, 'width:' ) && str_contains( $image, 'height:' ) ) {
		return true;
	}

	return false;
}
```

I tested this and it works, so the question is whether I'm on the right track or not. Should WP not add `loading="lazy"` in this scenario?

**Reply by jerclarke** [reply]
And just to clarify and be explicit: The underlying issue is that to work correctly, images MUST be embedded attachments currently. If I "replace" the image in the Site Editor with one from the media library, it also gets inserted with width/height that control the inline CSS, but the attachment logic of WP **also** inserts the `width=""` and `height=""` attributes to the tag, and thus the `loading="lazy"` logic considers it valid. 

But why can't I add an image in the theme directly, referencing a file in the theme dir? That's why we'd need an explicit block attribute for `width=` and `height=`, unless there's an unwritten rule that all images in block themes have to be attachments.

**Reply by jerclarke** [reply]
FWIW I realized I should just look and see how this is handled in the many patterns that include images by default. The result wasn't surprising I guess: None of them have any width/height applied to them, meaning after being inserted they are never given `loading` or `fetchpriority` values by WP. 

This in itself isn't the end of the world, those images are meant to be replaced, after all, but it does point to the fundamental issue I'm bringing up: **It's impossible to insert a pre-defined image into a pattern in a performant way.**

I can imagine many scenarios where a customized block theme (or child theme) would have special patterns defined for use on the site, where the images are a known quantity and thus are included in the theme itself, then referenced in the patterns. Currently that workflow will result in Lighthouse errors every time, unless the content editor manually replaces all the images every time. 

Seems like something worth fixing one way or another! We should either allow templates/parts/patterns to define a true `width=` and `height=` on `wp:image` (and `wp:cover` and any others), or get `wp_img_tag_add_loading_optimization_attrs()` to accept  `width:` and `height:` as triggers for `loading` and `fetchpriority`.
